### ISSUE TITLE
Why is WrappedStorageException an internal class?

### ISSUE THREAD

**Issue by shirhatti** [Azure/azure-storage-net#178]
I am currently consuming the Table Storage API in my project.

I do not follow why `WrappedStorageException` is an internal class. This  is currently how I'm consuming the API.

````csharp
try
{
    TableOperation ReplaceOperation = TableOperation.Replace(MyEntity);
    await myTable.ExecuteAsync(ReplaceOperation);
}
catch (Exception e)
{
    var innerException = e.InnerException as StorageException;
    if (innerException != null)
    {
        if (innerException.RequestInformation.HttpStatusCode == 412)
        {
            // Handle optimistic concurrency violation
        }
    }
    else
    {
        throw;
    }
}
````

 I'm not comfortable catching a generic exception and I cannot figure a more elegant way to do this in the code snippet above.

**Comment by GuardRex** [comment]
Why can't you catch it as a `StorageException`?
```c#
...
}
catch (StorageException e)
{
    if (e.RequestInformation.HttpStatusCode == 412)
    {
        // Handle optimistic concurrency violation
    }
}
```

**Comment by shirhatti** [comment]
Because the exception being thrown is NOT a `StorageException`, it is a `WrappedStorageException`

Here is the exception being thrown

````
"Microsoft.WindowsAzure.Storage.WrappedStorageException (0x8004119C): <?xml version=\"1.0\" encoding=\"utf-16\"?>
<!--An exception has occurred. For more information please deserialize this message via RequestResult.TranslateFromExceptionMessage.-->
<RequestResult>
  <HTTPStatusCode>412</HTTPStatusCode>
  <HttpStatusMessage>Precondition Failed</HttpStatusMessage>
  <TargetLocation>Primary</TargetLocation>
  <ServiceRequestID>63db9c03-0002-00d4-6c6d-ef583d000000</ServiceRequestID>
  <ContentMd5 />
  <Etag />
  <RequestDate>Mon, 14 Sep 2015 21:17:14 GMT</RequestDate>
  <StartTime>Tue, 15 Sep 2015 04:17:14 GMT</StartTime>
  <EndTime>Tue, 15 Sep 2015 04:17:14 GMT</EndTime>
  <Error>
    <Code>UpdateConditionNotSatisfied</Code>
    <Message>The update condition specified in the request was not satisfied.
RequestId:63db9c03-0002-00d4-6c6d-ef583d000000
Time:2015-09-15T04:17:19.1087238Z</Message>
  </Error>
  <ExceptionInfo>
    <Type>StorageException</Type>

    <HResult>-2146233088</HResult>
    <Message>Precondition Failed</Message>
    <Source>Microsoft.WindowsAzure.Storage</Source>
    <StackTrace>   at Microsoft.WindowsAzure.Storage.Core.Executor.Executor.&lt;ExecuteAsyncInternal&gt;d__4`1.MoveNext()</StackTrace>
  </ExceptionInfo>
</RequestResult> ---> Microsoft.WindowsAzure.Storage.StorageException: Precondition Failed
   at Microsoft.WindowsAzure.Storage.Core.Executor.Executor.<ExecuteAsyncInternal>d__4`1.MoveNext()
Request Information
RequestID:63db9c03-0002-00d4-6c6d-ef583d000000
RequestDate:Mon, 14 Sep 2015 21:17:14 GMT
StatusMessage:Precondition Failed
ErrorCode:UpdateConditionNotSatisfied
````

`StorageException` is the InnerException in the `WrappedStorageException`

**Comment by GuardRex** [comment]
What version are you using? Perhaps, they fixed it in the latest preview release. In `5.0.3-preview`, I see:
```c#
namespace Microsoft.WindowsAzure.Storage
{
    public class StorageException : Exception
    {
        public StorageException();
        public StorageException(string message);
        public StorageException(string message, Exception innerException);
        public StorageException(RequestResult res, string message, Exception inner);

        public RequestResult RequestInformation
        {
            get;
        }

        public static StorageException TranslateException(Exception ex, RequestResult reqResult);
        public static StorageException TranslateException(Exception ex, RequestResult reqResult, Func<Stream, StorageExtendedErrorInformation> parseError);
        public static StorageException TranslateException(Exception ex, RequestResult reqResult, Func<Stream, StorageExtendedErrorInformation> parseError, HttpResponseMessage response);
        public override string ToString();
    }
}
```

**Comment by GuardRex** [comment]
My bet is on "bug" ... try a newer client library if you can. It looks like this happened one or two times in the past. When it happened in one case, I think @nickharris told the person to upgrade their library.

**Comment by shirhatti** [comment]
I am currently using `5.0.1-preview`
I will try `5.0.3-preview` tomorrow and post an update here.

**Comment by pemari-msft** [comment]
What platform are you targeting? This was due to WindowsRT having limitations for what can be public.

**Comment by shirhatti** [comment]
I am targeting `dnx451` and `dnxcore50`

**Comment by pemari-msft** [comment]
Sure. We can make it so WrappedStorageException is public in an upcoming release. Thanks for the feedback!

**Comment by GuardRex** [comment]
I'm confused about exposing exceptions that way: Why aren't all storage client exceptions going to be caught as `StorageException`? I don't know when I should use one or the other in a `catch()`.

**Comment by sescandell** [comment]
Hi there,

I'm in the same situation as @shirhatti : I'm targeting `dnx451` & `dnxcore50` and using the `WindowsAzure.storage` in version `6.1.1-preview`.

`WrappedStorageException` is still internal. So we cannot catch (for example) duplicate entry on Azure Table Storage `Insert` Operation.

We can't catch `StorageException` because it's actually not the exception thrown (as you can see here: https://github.com/Azure/azure-storage-net/blob/40c04ce1dbbc0c569316da0e3fe47b9bff5d3042/Lib/WindowsRuntime/Core/Executor/Executor.cs#L119 )

Thanks for your help,

**Comment by jehine-MSFT** [comment]
Hello all,

We are planning on addressing this issue in a future release.

Thanks for the suggestion!

**Comment by poter134** [comment]
Any news on this? 409 Conflicts used to throw StorageExceptions.  I don't think WrappedStorageException should be made public (although it couldn't hurt), I'd rather the behaviour was changed back to how it used to be. 

We built logic around receiving this code that informed the user that the resource already existed.  Why the change?

**Comment by pemari-msft** [comment]
Hi, we are indeed making the change to only throw storage exceptions. This is an artifact of the winRT target and it is not new behavior. We will have it fixed in an upcoming major version release. Thanks!

**Comment by pemari-msft** [comment]
This is fixed in version 7.0 of our library. Please try it out!
